File: server/src/redflags.ts
/**
 * Deterministic Red-Flag Detection Engine
 *
 * SAFETY-CRITICAL: This module provides a hard safety net that runs
 * independently of any AI model. It uses pattern matching across
 * English, Hindi, and Marathi to detect life-threatening conditions.
 */

export interface RedFlagResult {
  hasRedFlags: boolean;
  detectedFlags: string[];
  forcedEsi: 1 | 2 | null;
}

// ── Clinical Pattern Definitions ──────────────────────────────────────────

interface ClinicalPattern {
  name: string;
  pattern: RegExp;
  /** If true, forces ESI 1 when matched */
  isEsi1: boolean;
}

const CLINICAL_PATTERNS: ClinicalPattern[] = [
  {
    name: 'Chest pain / Acute Coronary Syndrome indicators',
    pattern:
      /(chest pain|pressure in chest|heart attack|angina|tightness in chest|heaviness in chest|छाती में दर्द|सीने में दर्द|छाती पर दबाव|छातीत दुखणे|छातीत दाब|छाती दुखते|छातीत जडपणा)/i,
    isEsi1: true,
  },
  {
    name: 'Severe respiratory distress / Shortness of breath',
    pattern:
      /(difficulty breathing|shortness of breath|breathless|gasping|cannot breathe|can't breathe|suffocation|सांस लेने में तकलीफ|दम फूलना|सांस फूलना|श्वास घेण्यास त्रास|श्वास कोंडणे|दम लागणे)/i,
    isEsi1: true,
  },
  {
    name: 'Stroke / FAST signs',
    pattern:
      /(face droop|facial drooping|arm weakness|slurred speech|stroke|paralysis|sudden weakness|लकवा|फेशियल पाल्सी|मुंह टेढ़ा होना|बोलने में लड़खड़ाहट|अर्धांगवायू|चेहरा वाकडा|हात कमकुवत|बोलण्यात अडखळणे)/i,
    isEsi1: false,
  },
  {
    name: 'Severe hemorrhage / Uncontrolled bleeding',
    pattern:
      /(severe bleeding|uncontrolled bleed|hemorrhage|arterial bleed|blood spurting|gushing blood|भारी रक्तस्राव|अत्यधिक खून बहना|रक्त बहना रुक नहीं रहा|तीव्र रक्तस्त्राव|अति रक्तस्राव|रक्त थांबत नाही)/i,
    isEsi1: false,
  },
  {
    name: 'Unconsciousness / Unresponsive',
    pattern:
      /(unconscious|passed out|fainted|unresponsive|syncope|collapsed|blacked out|बेहोश|मूर्छित|होश खो बैठना|बेशुद्ध|चक्कर येऊन पडणे|जाणीव नसणे)/i,
    isEsi1: true,
  },
  {
    name: 'Seizure / Convulsions',
    pattern:
      /(seizure|convulsion|fits|epilepsy|twitching involuntarily|दौरे|झटके|मिरगी का दौरा|आकडी|फेफरे|फिट येणे)/i,
    isEsi1: false,
  },
];

// ── Vital Sign Thresholds ─────────────────────────────────────────────────

interface VitalCheck {
  name: string;
  /** Returns true if the vital is in dangerous range */
  check: (vitals: any) => boolean;
  /** If true, forces ESI 1 when triggered */
  isEsi1: boolean;
}

const VITAL_CHECKS: VitalCheck[] = [
  {
    name: 'Critical hypoxemia: SpO2 < 90%',
    check: (v) => {
      const spo2 = parseFloat(v?.spO2 ?? v?.spo2 ?? '');
      return !isNaN(spo2) && spo2 < 90;
    },
    isEsi1: true,
  },
  {
    name: 'Shock / Severe hypotension: Systolic BP < 90 mmHg',
    check: (v) => {
      const sbp = parseFloat(v?.systolicBp ?? v?.systolic_bp ?? v?.systolicBP ?? '');
      return !isNaN(sbp) && sbp < 90;
    },
    isEsi1: true,
  },
  {
    name: 'Dangerous tachycardia: Heart Rate > 130 bpm',
    check: (v) => {
      const hr = parseFloat(v?.heartRate ?? v?.heart_rate ?? v?.hr ?? '');
      return !isNaN(hr) && hr > 130;
    },
    isEsi1: false,
  },
];

// ── Main Detection Function ───────────────────────────────────────────────

/**
 * Scans free-text transcript and vital signs for life-threatening red flags.
 *
 * @param text  Combined patient transcript / chief complaint text
 * @param vitals  Optional vitals object with spO2, systolicBp, heartRate, etc.
 * @returns  RedFlagResult with detected flags and forced ESI level
 */
export function detectRedFlags(text: string, vitals?: any): RedFlagResult {
  const detectedFlags: string[] = [];
  let hasEsi1 = false;
  let hasEsi2 = false;

  // Check clinical patterns against text
  for (const pattern of CLINICAL_PATTERNS) {
    if (pattern.pattern.test(text)) {
      detectedFlags.push(pattern.name);
      if (pattern.isEsi1) {
        hasEsi1 = true;
      } else {
        hasEsi2 = true;
      }
    }
  }

  // Check vital sign thresholds
  if (vitals && typeof vitals === 'object') {
    for (const check of VITAL_CHECKS) {
      if (check.check(vitals)) {
        // Append actual value to the flag name for clarity
        let flagName = check.name;
        if (check.name.includes('SpO2')) {
          const val = vitals?.spO2 ?? vitals?.spo2;
          if (val !== undefined) flagName = `Critical hypoxemia: SpO2 ${val}% (< 90%)`;
        } else if (check.name.includes('Systolic BP')) {
          const val = vitals?.systolicBp ?? vitals?.systolic_bp ?? vitals?.systolicBP;
          if (val !== undefined) flagName = `Shock / Severe hypotension: Systolic BP ${val} mmHg (< 90 mmHg)`;
        } else if (check.name.includes('Heart Rate')) {
          const val = vitals?.heartRate ?? vitals?.heart_rate ?? vitals?.hr;
          if (val !== undefined) flagName = `Dangerous tachycardia: Heart Rate ${val} bpm (> 130 bpm)`;
        }
        detectedFlags.push(flagName);
        if (check.isEsi1) {
          hasEsi1 = true;
        } else {
          hasEsi2 = true;
        }
      }
    }
  }

  const hasRedFlags = detectedFlags.length > 0;
  let forcedEsi: 1 | 2 | null = null;
  if (hasEsi1) {
    forcedEsi = 1;
  } else if (hasEsi2) {
    forcedEsi = 2;
  }

  return { hasRedFlags, detectedFlags, forcedEsi };
}

File: backend/redflags.py
import re
from typing import Dict, List, Optional, Any, Tuple

CLINICAL_PATTERNS = [
    {
        "name": "Chest pain / Acute Coronary Syndrome indicators",
        "pattern": re.compile(
            r"(chest pain|pressure in chest|heart attack|angina|tightness in chest|heaviness in chest|"
            r"छाती में दर्द|सीने में दर्द|छाती पर दबाव|छातीत दुखणे|छातीत दाब|छाती दुखते|छातीत जडपणा)",
            re.IGNORECASE,
        ),
        "is_esi_1": True,
    },
    {
        "name": "Severe respiratory distress / Shortness of breath",
        "pattern": re.compile(
            r"(difficulty breathing|shortness of breath|breathless|gasping|cannot breathe|can't breathe|suffocation|"
            r"सांस लेने में तकलीफ|दम फूलना|सांस फूलना|श्वास घेण्यास त्रास|श्वास कोंडणे|दम लागणे)",
            re.IGNORECASE,
        ),
        "is_esi_1": True,
    },
    {
        "name": "Stroke / FAST signs",
        "pattern": re.compile(
            r"(face droop|facial drooping|arm weakness|slurred speech|stroke|paralysis|sudden weakness|"
            r"लकवा|फेशियल पाल्सी|मुंह टेढ़ा होना|बोलने में लड़खड़ाहट|अर्धांगवायू|चेहरा वाकडा|हात कमकुवत|बोलण्यात अडखळणे)",
            re.IGNORECASE,
        ),
        "is_esi_1": False,
    },
    {
        "name": "Severe hemorrhage / Uncontrolled bleeding",
        "pattern": re.compile(
            r"(severe bleeding|uncontrolled bleed|hemorrhage|arterial bleed|blood spurting|gushing blood|"
            r"भारी रक्तस्राव|अत्यधिक खून बहना|रक्त बहना रुक नहीं रहा|तीव्र रक्तस्त्राव|अति रक्तस्राव|रक्त थांबत नाही)",
            re.IGNORECASE,
        ),
        "is_esi_1": False,
    },
    {
        "name": "Unconsciousness / Unresponsive",
        "pattern": re.compile(
            r"(unconscious|passed out|fainted|unresponsive|syncope|collapsed|blacked out|"
            r"बेहोश|मूर्छित|होश खो बैठना|बेशुद्ध|चक्कर येऊन पडणे|जाणीव नसणे)",
            re.IGNORECASE,
        ),
        "is_esi_1": True,
    },
    {
        "name": "Seizure / Convulsions",
        "pattern": re.compile(
            r"(seizure|convulsion|fits|epilepsy|twitching involuntarily|"
            r"दौरे|झटके|मिरगी का दौरा|आकडी|फेफरे|फिट येणे)",
            re.IGNORECASE,
        ),
        "is_esi_1": False,
    },
]

def detect_red_flags(text: str, vitals: Optional[Dict[str, Any]] = None) -> Tuple[bool, List[str], Optional[int]]:
    """
    Deterministic clinical safety check.
    Returns: (has_red_flags, detected_flags, forced_esi: 1 | 2 | None)
    """
    detected_flags: List[str] = []
    has_esi_1_indicator = False

    # 1. Text pattern scan
    if text:
        for item in CLINICAL_PATTERNS:
            if item["pattern"].search(text):
                detected_flags.append(item["name"])
                if item["is_esi_1"]:
                    has_esi_1_indicator = True

    # 2. Vital signs threshold checks
    if vitals:
        # SpO2 < 90%
        spo2 = vitals.get("spO2") or vitals.get("spo2")
        if spo2 is not None:
            try:
                spo2_val = float(spo2)
                if 0 < spo2_val < 90:
                    detected_flags.append(f"Critical hypoxemia: SpO2 {spo2_val:.0f}% (< 90%)")
                    has_esi_1_indicator = True
            except (ValueError, TypeError):
                pass

        # Systolic BP < 90 mmHg (Shock)
        sys_bp = vitals.get("systolicBp") or vitals.get("systolic_bp")
        if sys_bp is not None:
            try:
                sys_val = float(sys_bp)
                if 0 < sys_val < 90:
                    detected_flags.append(f"Shock / Severe hypotension: Systolic BP {sys_val:.0f} mmHg (< 90 mmHg)")
                    has_esi_1_indicator = True
            except (ValueError, TypeError):
                pass

        # Heart Rate > 130 bpm
        hr = vitals.get("heartRate") or vitals.get("heart_rate")
        if hr is not None:
            try:
                hr_val = float(hr)
                if hr_val > 130:
                    detected_flags.append(f"Dangerous tachycardia: Heart Rate {hr_val:.0f} bpm (> 130 bpm)")
                    # Tachycardia alone is ESI 2 unless combined with hypotension/chest pain
            except (ValueError, TypeError):
                pass

    has_flags = len(detected_flags) > 0
    forced_esi: Optional[int] = None
    if has_flags:
        forced_esi = 1 if has_esi_1_indicator else 2

    return has_flags, detected_flags, forced_esi
